system/api-auth.go

```diff
@@ -0,0 +1,131 @@
+	}
+}
+
+func (application *Application) validateTicketOwnership(authHeader string) (multiSigAddress, authValidationFailureReason string) {
+	timestamp, timestampSignature, ticketHash := extractTicketAuthParams(strings.TrimPrefix(authHeader, "TicketAuth "))
+	if timestamp == "" || timestampSignature == "" || ticketHash == "" {
+		authValidationFailureReason = fmt.Sprintf("invalid ticket auth header value %s", authHeader)
+		return
+	}
+
+	// Ensure that this signature had not been used in a previous authentication attempt.
+	if application.ProcessedTicketChallenges.ContainsChallenge(timestampSignature) {
+		authValidationFailureReason = fmt.Sprintf("disallowed reuse of ticket auth signature %v", timestampSignature)
+		return
+	}
+
+	authTimestamp, err := strconv.Atoi(timestamp)
+	if err != nil {
+		authValidationFailureReason = fmt.Sprintf("invalid ticket auth timestamp value %v", timestamp)
+		return
+	}
+
+	// Ensure that the auth timestamp is not more than
+	// the permitted number of seconds into the past and future.
+	timestampDelta := time.Now().Unix() - int64(authTimestamp)
+	if math.Abs(float64(timestampDelta)) > maxTicketChallengeAge {
+		authValidationFailureReason = fmt.Sprintf("expired or invalid ticket auth timestamp value %v",
+			timestamp)
+		return
+	}
+
+	// confirm that the timestamp signature is a valid base64 string
+	decodedSignature, err := base64.StdEncoding.DecodeString(timestampSignature)
+	if err != nil {
+		authValidationFailureReason = fmt.Sprintf("invalid ticket auth signature %s", timestampSignature)
+		return
+	}
+
+	// Mark this timestamp signature as used to prevent subsequent reuse.
+	application.ProcessedTicketChallenges.AddChallenge(timestampSignature, maxTicketChallengeAge)
+
+	// todo check if ticket belongs to this vsp
+
+	// get user wallet address using ticket hash
+	// todo: may be better to maintain a memory map of tickets-userWalletAddresses
+	ticketInfo, err := application.StakepooldConnMan.GetTicketInfo(ticketHash)
+	if err != nil {
+		authValidationFailureReason = fmt.Sprintf("ticket auth, get ticket info failed: %v", err)
+		return
+	}
+
+	// check if timestamp signature checks out against address
+	addr, err := dcrutil.DecodeAddress(ticketInfo.OwnerFeeAddress)
+	if err != nil {
+		authValidationFailureReason = fmt.Sprintf("ticket auth, unexpected decode address error: %v", err)
+		return
+	}
+
+	valid, err := wallet.VerifyMessage(timestamp, addr, decodedSignature)
+	if err != nil {
+		authValidationFailureReason = fmt.Sprintf("error validating timestamp signature for ticket auth %v", err)
+		return
+	}
+
+	if valid {
+		multiSigAddress = ticketInfo.MultiSigAddress
+	} else {
+		authValidationFailureReason = "invalid timestamp signature, timestamp was not signed using " +
+			"the ticket owner's reward address"
+	}
+	return
+}
+
+func extractTicketAuthParams(authHeader string) (timestampMessage, timestampSignature, ticketHash string) {
+	authParams := strings.Split(authHeader, ",")
```

Comment:
The authorization header format was desigend to match the http authorization specification defined in [RFC2617](http://www.ietf.org/rfc/rfc2617.txt). [AWS Signature Version 4](https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-auth-using-authorization-header.html) is an example.

Reply:
Will this perhaps be better?
`authParams, _ := url.ParseQuery(strings.ReplaceAll(authHeader, ",", "&"))`

It's similar string parsing that goes on under the hood as well, so don't really see an advantage to this approach. What do you think?

Reply:
If it's not possible to change on the consumer's side, then you can ignore this.  I just thought it might be cleaner, but it seems commas are the standard.  Maybe there's a good reason, I don't know though. If you know please tell me...